File: lib/scrapers/seriesScraper.js
const axios = require('axios');
const cheerio = require('cheerio');
const { BASE_URL, fetchPage, extractIdFromUrl, cleanText, makeAbsoluteUrl, HEADERS } = require('../utils');

/**
 * Scrape del listado de series desde /series-online/
 */
async function scrapeSeriesList(page = 1) {
  const url = page === 1
    ? `${BASE_URL}/series-online/`
    : `${BASE_URL}/series-online/page/${page}/`;

  const $ = await fetchPage(url);
  if (!$) return [];

  const seriesList = [];

  $('.TPost.B, article.TPost.B').each((i, item) => {
    try {
      const $item = $(item);
      let link = $item.find('a[href*="/serie/"]').first();
      if (!link.length) link = $item.find('a').first();

      const titleElem = $item.find('.Title, a.Title').first();
      const img = $item.find('.Image img, figure img').first();
      const yearElem = $item.find('.Qlty').first();
      const infoElem = $item.find('.Info').first();

      const href = link.attr('href');
      if (href) {
        const url = makeAbsoluteUrl(href);

        let year = null;
        if (yearElem.length) {
          year = cleanText(yearElem.text());
        } else if (infoElem.length) {
          const infoText = cleanText(infoElem.text());
          const yearMatch = infoText.match(/(\d{4})/);
          if (yearMatch) year = yearMatch[1];
        }

        seriesList.push({
          id: extractIdFromUrl(url),
          title: cleanText(titleElem.text()) || cleanText(link.text()),
          url,
          image: img.length ? makeAbsoluteUrl(img.attr('src') || img.attr('data-src') || '') : null,
          year,
          rating: null,
        });
      }
    } catch (error) {
      console.error('Error parsing series:', error.message);
    }
  });

  return seriesList;
}

/**
 * Scrape del detalle de una serie incluyendo temporadas y episodios
 */
async function scrapeSeriesDetail(seriesId) {
  const seriesUrl = `${BASE_URL}/serie/${seriesId}/`;
  let $ = await fetchPage(seriesUrl);

  if (!$) {
    $ = await fetchPage(`${BASE_URL}/serie/${seriesId}`);
    if (!$) return null;
  }

  try {
    const titleElem = $('h1.Title, .Title, h1').first();
    const title = cleanText(titleElem.text()) || seriesId;

    const imgElem = $('.TPost img, .Image img, article img, figure img').first();
    const image = imgElem.length ? makeAbsoluteUrl(imgElem.attr('src') || imgElem.attr('data-src') || '') : null;

    const descriptionElem = $('.Description, p.Description, .TPMvCn p').first();
    const description = descriptionElem.length ? cleanText(descriptionElem.text()) : null;

    const yearElem = $('.Qlty, .Info .Qlty').first();
    let year = null;
    if (yearElem.length) {
      const yearText = cleanText(yearElem.text());
      const yearMatch = yearText.match(/(\d{4})/);
      if (yearMatch) year = yearMatch[1];
    }

    const genres = [];
    $('a[href*="/genero/"]').slice(0, 5).each((i, elem) => {
      const genreText = cleanText($(elem).text());
      if (genreText && !genres.includes(genreText)) {
        genres.push(genreText);
      }
    });

    const cast = [];
    const castRegex = /Actores?:/i;
    $('*').filter(function() {
      return $(this).text().match(castRegex);
    }).first().find('a').slice(0, 10).each((i, elem) => {
      const actorName = cleanText($(elem).text());
      if (actorName) cast.push(actorName);
    });

    const seasons = await scrapeSeasonsFromLinks($, seriesId);

    return {
      id: seriesId,
      title,
      url: seriesUrl,
      image,
      description,
      year,
      rating: null,
      genres,
      cast,
      seasons,
    };
  } catch (error) {
    console.error('Error parsing series detail:', error.message);
    return null;
  }
}

/**
 * Extrae las temporadas desde los enlaces /temporada/
 */
async function scrapeSeasonsFromLinks($, seriesId) {
  const seasons = [];
  const seasonLinks = $('a[href*="/temporada/"]');

  if (seasonLinks.length === 0) {
    // Si no hay enlaces a temporadas, buscar episodios directamente
    const episodeLinks = $('a[href*="/episodio/"]');
    const episodes = [];

    episodeLinks.each((i, link) => {
      try {
        const url = makeAbsoluteUrl($(link).attr('href'));
        const epId = extractIdFromUrl(url);
        const match = epId.match(/-(\d+)x(\d+)/);
        const epNum = match ? parseInt(match[2]) : episodes.length + 1;

        episodes.push({
          number: epNum,
          title: cleanText($(link).text()) || `Episodio ${epNum}`,
          url,
          image: null,
          servers: [],
        });
      } catch (error) {
        console.error('Error parsing episode link:', error.message);
      }
    });

    if (episodes.length > 0) {
      seasons.push({ number: 1, episodes });
    }

    return seasons;
  }

  const seasonDict = {};

  for (let i = 0; i < seasonLinks.length; i++) {
    try {
      const link = seasonLinks[i];
      const seasonUrl = makeAbsoluteUrl($(link).attr('href'));
      const seasonSlug = extractIdFromUrl(seasonUrl);
      const seasonMatch = seasonSlug.match(/-(\d+)\/?$/);

      if (!seasonMatch) continue;
      const seasonNum = parseInt(seasonMatch[1]);

      if (seasonDict[seasonNum]) continue;

      const episodes = await scrapeSeasonEpisodes(seasonUrl);
      if (episodes.length > 0) {
        seasonDict[seasonNum] = { number: seasonNum, episodes };
      }
    } catch (error) {
      console.error('Error parsing season link:', error.message);
    }
  }

  const sortedSeasons = Object.keys(seasonDict)
    .sort((a, b) => parseInt(a) - parseInt(b))
    .map(num => seasonDict[num]);

  return sortedSeasons;
}

/**
 * Extrae los episodios de una temporada específica
 */
async function scrapeSeasonEpisodes(seasonUrl) {
  const $ = await fetchPage(seasonUrl);
  if (!$) return [];

  const episodes = [];
  const episodeLinks = $('a[href*="/episodio/"]');

  episodeLinks.each((i, link) => {
    try {
      const url = makeAbsoluteUrl($(link).attr('href'));
      const epId = extractIdFromUrl(url);
      const match = epId.match(/-\d+x(\d+)/);
      const epNum = match ? parseInt(match[1]) : episodes.length + 1;

      const parent = $(link).closest('.TPost, article, li');
      let img = null;
      let titleText = cleanText($(link).text());

      if (parent.length) {
        const imgElem = parent.find('img').first();
        if (imgElem.length) {
          img = makeAbsoluteUrl(imgElem.attr('src') || imgElem.attr('data-src') || '');
        }

        const titleElem = parent.find('.Title').first();
        if (titleElem.length) {
          titleText = cleanText(titleElem.text());
        }
      }

      const episode = {
        number: epNum,
        title: titleText || `Episodio ${epNum}`,
        url,
        image: img,
        servers: [],
      };

      if (!episodes.find(e => e.number === epNum)) {
        episodes.push(episode);
      }
    } catch (error) {
      console.error('Error parsing episode:', error.message);
    }
  });

  episodes.sort((a, b) => a.number - b.number);
  return episodes;
}

/**
 * Extrae los servidores de streaming de un episodio
 */
async function scrapeEpisodeServers(episodeUrl) {
  const $ = await fetchPage(episodeUrl);
  if (!$) return [];

  const servers = [];
  const seenUrls = new Set();

  const addServer = (name, url, quality = null, language = null) => {
    if (!url) return;
    const fullUrl = url.startsWith('http') ? url : makeAbsoluteUrl(url);
    if (seenUrls.has(fullUrl)) return;

    seenUrls.add(fullUrl);
    servers.push({ name, url: fullUrl, quality, language });
  };

  const resolveServerUrl = $elem => {
    const possibleAttrs = [
      'data-video',
      'data-src',
      'data-url',
      'data-embed',
      'data-link',
      'data-file',
      'href',
      'value',
    ];

    for (const attr of possibleAttrs) {
      const value = $elem.attr(attr);
      if (value) return value;
    }

    const iframe = $elem.find('iframe').first();
    if (iframe.length) {
      return iframe.attr('src') || iframe.attr('data-src') || '';
    }

    return '';
  };

  const extractOptionData = $elem => {
    const labelElem = $elem.find('.server-name, .title, span').first();
    const label = cleanText(labelElem.text()) || cleanText($elem.text()) || 'Servidor';
    const language = cleanText($elem.attr('data-lang') || $elem.attr('data-language') || '');
    const quality = cleanText($elem.attr('data-quality') || $elem.attr('data-res') || '');

    return {
      label,
      language: language || null,
      quality: quality || null,
      embed: resolveServerUrl($elem),
      postId: $elem.attr('data-post') || null,
      nume: $elem.attr('data-nume') || null,
      type: $elem.attr('data-type') || null,
    };
  };

  const resolveAjaxIframe = async ({ postId, nume, type }) => {
    if (!postId || !nume) return '';

    try {
      const ajaxUrl = `${BASE_URL}/wp-admin/admin-ajax.php`;
      const payload = new URLSearchParams({
        action: 'doo_player_ajax',
        post: postId,
        nume,
        type: type || 'iframe',
      });

      const response = await axios.post(ajaxUrl, payload.toString(), {
        headers: {
          ...HEADERS,
          'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
          Referer: episodeUrl,
        },
      });

      const iframeHtml = response.data || '';
      const $iframe = cheerio.load(iframeHtml);
      const iframe = $iframe('iframe').first();

      if (iframe.length) {
        return iframe.attr('src') || iframe.attr('data-src') || '';
      }

      const match = String(iframeHtml).match(/src=["']([^"']+)["']/i);
      return match ? match[1] : '';
    } catch (error) {
      console.error('Error resolving player iframe:', error.message);
      return '';
    }
  };

  // Opciones de player agrupadas por idioma (ej. Latino, Castellano, Subtitulado)
  const playerOptionElems = $(
    '#playeroptionsul li, .playeroptions li, li.dooplay_player_option, li[data-post][data-nume]'
  );

  for (let i = 0; i < playerOptionElems.length; i++) {
    try {
      const optionData = extractOptionData($(playerOptionElems[i]));
      let finalUrl = optionData.embed;

      if (!finalUrl) {
        finalUrl = await resolveAjaxIframe(optionData);
      }

      addServer(optionData.label, finalUrl, optionData.quality, optionData.language);
    } catch (error) {
      console.error('Error parsing player option:', error.message);
    }
  }

  // Algunas páginas usan selects para alternar idioma/servidor
  $('select option[data-post], select option[data-video], select option[data-embed]').each((i, elem) => {
    try {
      const optionData = extractOptionData($(elem));
      let finalUrl = optionData.embed;

      if (!finalUrl) {
        finalUrl = await resolveAjaxIframe(optionData);
      }

      addServer(optionData.label, finalUrl, optionData.quality, optionData.language);
    } catch (error) {
      console.error('Error parsing select option:', error.message);
    }
  });

  // Fallback: elementos genéricos con data-video/embed y sin ajax
  $(
    '.server, .player-option, .option, li[data-video], .play-box-iframe, li[data-embed], button[data-video], button[data-embed]'
  ).each((i, elem) => {
    try {
      const optionData = extractOptionData($(elem));
      addServer(optionData.label, optionData.embed, optionData.quality, optionData.language);
    } catch (error) {
      console.error('Error parsing server element:', error.message);
    }
  });

  // Algunos episodios incluyen la lista de servidores en scripts inline
  $('script').each((i, script) => {
    try {
      const content = $(script).html() || '';
      const match = content.match(/sources?\s*[:=]\s*(\[[^\]]+\])/i);

      if (match && match[1]) {
        const normalized = match[1]
          .replace(/(\w+):/g, '"$1":')
          .replace(/'([^']*)'/g, '"$1"');

        const parsed = JSON.parse(normalized);

        parsed.forEach(source => {
          const file = source.file || source.src || source.url;
          const label = source.label || source.title || 'Servidor';
          const quality = source.res || source.quality || null;
          const language = source.lang || source.language || null;
          addServer(label, file, quality, language);
        });
      }
    } catch (error) {
      console.error('Error parsing inline servers:', error.message);
    }
  });

  return servers;
}

module.exports = {
  scrapeSeriesList,
  scrapeSeriesDetail,
  scrapeEpisodeServers,
};

File: scrapers/series_scraper.py
from typing import Optional, List
from schemas import SeriesBase, SeriesDetail, Season, Episode, Server
from utils import fetch_page, extract_id_from_url, clean_text, make_absolute_url
from config import BASE_URL
import re


async def scrape_series_list(page: int = 1) -> List[SeriesBase]:
    """
    Scrape del listado de series
    """
    url = f"{BASE_URL}/series" if page == 1 else f"{BASE_URL}/series/page/{page}"
    soup = await fetch_page(url)

    if not soup:
        return []

    series_list = []
    items = soup.select(".items .item, .series-list .item, article.item, .tvshows article")

    for item in items:
        try:
            link = item.select_one("a")
            img = item.select_one("img")
            title_elem = item.select_one(".title, h2, h3, .data h3")
            year_elem = item.select_one(".year, .date")
            rating_elem = item.select_one(".rating, .imdb")

            if link and link.get("href"):
                url = make_absolute_url(link.get("href"))
                series = SeriesBase(
                    id=extract_id_from_url(url),
                    title=clean_text(title_elem.get_text() if title_elem else link.get("title", "")),
                    url=url,
                    image=make_absolute_url(img.get("src") or img.get("data-src", "")) if img else None,
                    year=clean_text(year_elem.get_text()) if year_elem else None,
                    rating=clean_text(rating_elem.get_text()) if rating_elem else None,
                )
                series_list.append(series)
        except Exception as e:
            print(f"Error parsing series: {e}")
            continue

    return series_list


async def scrape_series_detail(series_id: str) -> Optional[SeriesDetail]:
    """
    Scrape del detalle de una serie incluyendo temporadas y episodios
    """
    # Intentar diferentes formatos de URL
    possible_urls = [
        f"{BASE_URL}/series/{series_id}",
        f"{BASE_URL}/serie/{series_id}",
        f"{BASE_URL}/tv/{series_id}",
    ]

    soup = None
    series_url = ""

    for url in possible_urls:
        soup = await fetch_page(url)
        if soup:
            series_url = url
            break

    if not soup:
        return None

    try:
        # Información básica
        title_elem = soup.select_one("h1, .title, .data h1")
        title = clean_text(title_elem.get_text()) if title_elem else series_id

        img_elem = soup.select_one(".poster img, .thumbnail img, article img")
        image = make_absolute_url(img_elem.get("src") or img_elem.get("data-src", "")) if img_elem else None

        description_elem = soup.select_one(".description, .wp-content, .summary p, .texto")
        description = clean_text(description_elem.get_text()) if description_elem else None

        year_elem = soup.select_one(".year, .date, .release-date")
        year = clean_text(year_elem.get_text()) if year_elem else None

        rating_elem = soup.select_one(".rating, .imdb, .dt_rating_vgs")
        rating = clean_text(rating_elem.get_text()) if rating_elem else None

        # Géneros
        genres = []
        genre_elems = soup.select(".genres a, .genre a, .sgeneros a")
        for genre in genre_elems:
            genres.append(clean_text(genre.get_text()))

        # Cast
        cast = []
        cast_elems = soup.select(".cast a, .actor a, .persons a")
        for actor in cast_elems[:10]:
            cast.append(clean_text(actor.get_text()))

        # Temporadas y episodios
        seasons = await scrape_seasons(soup, series_url, series_id)

        series_detail = SeriesDetail(
            id=series_id,
            title=title,
            url=series_url,
            image=image,
            description=description,
            year=year,
            rating=rating,
            genres=genres,
            cast=cast,
            seasons=seasons,
        )

        return series_detail

    except Exception as e:
        print(f"Error parsing series detail: {e}")
        return None


async def scrape_seasons(soup, series_url: str, series_id: str) -> List[Season]:
    """
    Extrae las temporadas y episodios de una serie
    """
    seasons = []

    # Buscar selectores de temporadas
    season_selectors = soup.select(".season, .se-c, #seasons .se-q")

    if not season_selectors:
        # Si no hay selector de temporadas, buscar episodios directamente
        season_selectors = [soup]

    for idx, season_elem in enumerate(season_selectors, 1):
        try:
            # Obtener número de temporada
            season_num_elem = season_elem.select_one(".se-t, .title, .season-number")
            if season_num_elem:
                season_text = season_num_elem.get_text()
                season_match = re.search(r'(\d+)', season_text)
                season_num = int(season_match.group(1)) if season_match else idx
            else:
                season_num = idx

            # Extraer episodios
            episodes = []
            episode_elems = season_elem.select(".episode, .episodio, .se-a li, .eps-item")

            for ep_elem in episode_elems:
                try:
                    ep_link = ep_elem.select_one("a")
                    if not ep_link or not ep_link.get("href"):
                        continue

                    ep_url = make_absolute_url(ep_link.get("href"))
                    ep_img = ep_elem.select_one("img")
                    ep_title_elem = ep_elem.select_one(".title, .episodiotitle, .epl-title")

                    # Extraer número de episodio
                    ep_num_elem = ep_elem.select_one(".episode-number, .numerando, .epl-num")
                    if ep_num_elem:
                        ep_text = ep_num_elem.get_text()
                        ep_match = re.search(r'(\d+)', ep_text)
                        ep_num = int(ep_match.group(1)) if ep_match else len(episodes) + 1
                    else:
                        ep_num = len(episodes) + 1

                    episode = Episode(
                        number=ep_num,
                        title=clean_text(ep_title_elem.get_text()) if ep_title_elem else f"Episodio {ep_num}",
                        url=ep_url,
                        image=make_absolute_url(ep_img.get("src") or ep_img.get("data-src", "")) if ep_img else None,
                        servers=[]  # Los servidores se cargan al consultar el episodio específico
                    )
                    episodes.append(episode)
                except Exception as e:
                    print(f"Error parsing episode: {e}")
                    continue

            if episodes:
                season = Season(number=season_num, episodes=episodes)
                seasons.append(season)

        except Exception as e:
            print(f"Error parsing season: {e}")
            continue

    return seasons


async def scrape_episode_servers(episode_url: str) -> List[Server]:
    """
    Extrae los servidores de streaming de un episodio
    """
    soup = await fetch_page(episode_url)
    if not soup:
        return []

    servers = []

    # Selectores comunes para servidores
    server_elems = soup.select(".server, .player-option, .option, li[data-video], .play-box-iframe")

    for server_elem in server_elems:
        try:
            name_elem = server_elem.select_one(".server-name, .title, span")
            name = clean_text(name_elem.get_text()) if name_elem else "Servidor"

            # URL del servidor
            server_url = (
                server_elem.get("data-video") or
                server_elem.get("data-src") or
                server_elem.get("data-url") or
                ""
            )

            # Si no está en atributo, buscar en iframe
            if not server_url:
                iframe = server_elem.select_one("iframe")
                if iframe:
                    server_url = iframe.get("src") or iframe.get("data-src", "")

            if server_url:
                quality_elem = server_elem.select_one(".quality, .qlty")
                quality = clean_text(quality_elem.get_text()) if quality_elem else None

                server = Server(
                    name=name,
                    url=make_absolute_url(server_url) if not server_url.startswith("http") else server_url,
                    quality=quality
                )
                servers.append(server)
        except Exception as e:
            print(f"Error parsing server: {e}")
            continue

    return servers
